Interactivity API bug: Fix fatal error when trying to use a stdClass as state-context.

cbravobernal opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket:  https://core.trac.wordpress.org/ticket/61039

Updates https://github.com/WordPress/wordpress-develop/pull/6581 with an unit test.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

narenin committed:
Fixed : Return value of  isn't used

narenin committed:
Fixed - Interactivity API issue for Some property access does not work well in server directives

narenin committed:
Removed unrelated changes

narendraie committed:
Implemented Suggestions

cbravobernal committed:
Add unit test

cbravobernal committed:
Remove not needed obj

sirreal commented on tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php:
```diff
@@ -785,6 +788,9 @@ public function test_evaluate_value() {
 		$this->assertEquals( 'otherPlugin-context', $result );
+
+		$result = $this->evaluate( 'state.obj.prop' );
+		$this->assertEquals( 'property', $result );
```
```suggestion
		$this->assertSame( 'property', $result );
```

cbravobernal committed:
Additional check

sirreal commented on tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php:
```diff
@@ -748,14 +748,17 @@ public function test_process_directives_does_not_change_inner_html_in_math() {
-	private function evaluate( $directive_value ) {
-		$generate_state = function ( $name ) {
+	private function evaluate( $directive_value, ) {
+		$obj            = new stdClass();
+		$obj->prop      = 'property';
+		$generate_state = function ( $name, $obj = null ) {
```
It seems like this could be simplified and follow the same pattern as the rest of the state and statically set `$obj->property = $name` instead of taking an `$obj` parameter.

sirreal commented on tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php:
```diff
@@ -748,14 +748,17 @@ public function test_process_directives_does_not_change_inner_html_in_math() {
 	 */
-	private function evaluate( $directive_value ) {
-		$generate_state = function ( $name ) {
+	private function evaluate( $directive_value, ) {
```
Trailing comma may be a problem in some supported PHP versions.
```suggestion
	private function evaluate( $directive_value ) {
```

cbravobernal replied:
it is being indeed, thanks! I was getting crazy

sirreal reviewed (approved):
This looks good in general, thanks both!

cbravobernal committed:
Remove trailing comma

cbravobernal committed:
Address suggestion

michalczaplinski commented:
It looks good to me, too! Nice work 👍

sirreal commented on tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php:
```diff
@@ -750,16 +750,20 @@ public function test_process_directives_does_not_change_inner_html_in_math() {
 				'key'    => $name,
 				'nested' => array( 'key' => $name . '-nested' ),
+				'obj'    => $obj,
 			);
```
We can also test array access:
```suggestion
				'arrAccess' => new class() implements ArrayAccess {
					public function offsetExists($offset): bool { return true; }
					public function offsetGet($offset): mixed { return $offset; }
					public function offsetSet($offset, $value): void {}
					public function offsetUnset($offset): void {}
				},
			);
```

sirreal commented on tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php:
```diff
@@ -785,6 +789,9 @@ public function test_evaluate_value() {
 		$this->assertEquals( 'otherPlugin-context', $result );
+
+		$result = $this->evaluate( 'state.obj.prop' );
+		$this->assertSame( 'myPlugin-state', $result );
```
This is interesting, I don't know what result we'll get here. I assume it's a string `1`. I'm not sure how the path is parsed and if it accounts for numeric access or coercion happens so that numeric string paths access arrays:

```suggestion
		$this->assertSame( 'myPlugin-state', $result );

		$result = $this->evaluate( 'state.arrAccess.1' );
		$this->assertSame( '1', $result );
```

cbravobernal replied:
In this case, `state.arrAccess.1`, 1 is treated first a string. You can operate, in example 
```
public function offsetGet( $offset ): mixed {
						return 2 * $offset; }
```
And will return 2 as a number.

What would be the best approach here?

sirreal replied:
I think it's fine as it. It's just interesting that `"1"` is used to access numeric arrays.

sirreal reviewed (commented):
Shall we test `ArrayAccess` too?

cbravobernal committed:
Add array access test

cbravobernal committed:
removed not compatible mixed type in > 8

cbravobernal committed:
Try return type will change

cbravobernal commented on tests/phpunit/tests/interactivity-api/wpInteractivityAPI.php:
```diff
@@ -750,16 +750,29 @@ public function test_process_directives_does_not_change_inner_html_in_math() {
+				'arrAccess' => new class() implements ArrayAccess {
+					public function offsetExists( $offset ): bool {
+						return true; }
+					#[\ReturnTypeWillChange]
```
@sirreal that test implies adding a this line in order to be compatible with PHP 7.2

It seems that the mixed typing is only available in PHP 8 and above.

Should we keep this test then?

gziolo replied:
Can you omit the `bool` type hint as the return value?

sirreal replied:
I tested this for the class body and it works (without errors, deprecations, or warnings) on php 8.3 and 7.2, let's use this:
```php
					public function offsetExists( $offset ): bool { return true; }
					#[\ReturnTypeWillChange]
					public function offsetGet( $offset ) { return $offset; }
					public function offsetSet( $offset, $value ): void {}
					public function offsetUnset( $offset ): void {}
```

> Can you omit the `bool` type hint as the return value?

That triggers a deprecation notice in recent versions (8.3).

sirreal replied:
I'd like to keep the test, if we support the behavior a test is a good thing 🙂

cbravobernal replied:
> Can you omit the `bool` type hint as the return value?

Nope, it will trigger:

`Fatal error: During inheritance of ArrayAccess: Uncaught Return type of ArrayAccess@anonymous::offsetExists($offset) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice`

gziolo replied:
This is what I meant based on the usage in other places:
<img width="1219" alt="Screenshot 2024-06-03 at 14 09 44" src="https://github.com/WordPress/wordpress-develop/assets/699132/5a0d9578-4e28-4434-a576-40370ff86353">

gziolo commented:
@cbravobernal, is this patch ready to land? https://github.com/WordPress/wordpress-develop/pull/6672#discussion_r1622282621 – any actions to take here?

cbravobernal committed:
Implement returntypewillchange similar to other places

cbravobernal commented:
> @cbravobernal[<img alt="" width="12" height="12" src="chrome-extension://hgomfjikakokcbkjlfgodhklifiplmpg/images/wp-logo.png">](https://profiles.wordpress.org/cbravobernal), is this patch ready to land? [#6672 (comment)](https://github.com/WordPress/wordpress-develop/pull/6672#discussion_r1622282621) – any actions to take here?

I removed the bool type and added `#[\ReturnTypeWillChange]` at the beginning of the function like you mentioned to keep consistency with other parts of the code. Let's see the tests.

cbravobernal committed:
Move function to the beginning of offset

cbravobernal commented:
All green so ready to land!

gziolo commented:
Committed with https://core.trac.wordpress.org/changeset/58320.
